app/models/post.rb

```diff
@@ -55,11 +55,30 @@ class Post < ApplicationRecord
   has_many :post_follows, dependent: :destroy
   has_many :comments, dependent: :destroy
 
+  scope :sfw, -> { where(nsfw: false) }
   scope :in_group, ->(group) { where(target_group: group) }
   scope :visible_for, ->(user) {
-    joins(:target_group).merge(Group.visible_for(user))
-                        .or(joins(:target_group)
-                        .where(target_group_id: nil))
+    scope = user && !user.sfw_filter? ? all : sfw
+
+    left_join = <<~SQL
+      LEFT OUTER JOIN groups g
+      ON g.id = target_group_id
+    SQL
+    group_visible = <<~SQL
+      posts.target_group_id IS NULL
+      OR (g.privacy = 0 OR g.privacy = 2)
+    SQL
+
+    return scope.joins(left_join).distinct.where(group_visible) unless user
+
+    scope.joins(left_join).distinct.where(<<~SQL)
+      #{group_visible}
+      OR g.id IN (
+        SELECT group_members.group_id
+        FROM group_members
+        WHERE group_members.user_id = #{user.id}
+      )
+    SQL
```

Comment:
Hmm, was there an issue which prevented implementing these within the existing `.or`?

Reply:
yup, iirc you can't use `.or` with joins unless they're nearly exactly the same. merging with `Group.visible_for` breaks that.

Reply:
Yeah I think you gotta have identical joins on either side of the `or`, hence the `joins(:target_group)` on either side previously

Reply:
Instead of `.merge(Group.visible_for)` could we do `where(target_group_id: Group.visible_for`? I think that'll generate an `IN` with a subquery

Reply:
i'll give it a try